src/directives/badge/badge.directive.ts

```diff
@@ -0,0 +1,170 @@
+        this._size = this._determineSize(size);
+    }
+
+    // shows + for number, ellipsis for text
+    public maxCharacters: number;
+
+    private _backgroundColor: ThemeColor;
+    private _fontColor: ThemeColor;
+
+    private _darkColor: ThemeColor;
+    private _lightColor: ThemeColor;
+    private _isNumber: boolean;
+    private _text: string;
+    private _displayText: string;
+    private _ariaLabel: string;
+    private _vPos: BadgeVerticalPosition;
+    private _hPos: BadgeHorizontalPosition;
+    private _size: BadgeSize;
+    private _overlap: boolean;
+    private _maxLength: number;
+    private suffix: string;
+
+    constructor(
+        private readonly _elementRef: ElementRef,
+        private readonly _renderer: Renderer,
+        private readonly _colorService: ColorService,
+        private readonly _contrastService: ContrastService
+    ) {
+    }
+
+    public ngOnInit(): void {
+        this._darkColor = ThemeColor.parse('#000');
+        this._lightColor = ThemeColor.parse('#fff');
+        this._vPos = BadgeVerticalPosition.Top;
+        this._hPos = BadgeHorizontalPosition.Right;
+        this._text = '';
+        this._ariaLabel = '';
+        this._maxLength = null;
+        this._overlap = true;
+        this._isNumber = false;
+    }
+
+    public ngOnChanges(changes: SimpleChanges): void {
+        // get display friendly version of text based on max length and type of val;
+        if (changes.text) {
+            let finalText: string = changes.badgeText.currentValue || '';
+
+            if (finalText.length) {
+                if (this._maxLength && finalText.length > this._maxLength) {
+                    if (this._isNumber) {
+                        finalText = '';
+
+                        for (let i = 0; i < this._maxLength; i++) {
+                            finalText += '9';
+                        }
+
+                        finalText += '+';
+                    } else {
+                        if (changes.badgeText.currentValue.length > this._maxLength) {
+                            finalText = `${ changes.text.currentValue.substr(0, this._maxLength - 1) }&hellip;`;
+                        }
+                    }
+                }
+            }
+
+            this._displayText = finalText;
+        }
+
+        // if text is only set, set aria label from it
+        if (changes.badgeText && !changes.badgeAriaLabel && !this._ariaLabel) {
+            this._ariaLabel = changes.badgeText.currentValue;
+        }
+    }
+
+    private _determineContrastColor(): ThemeColor {
+        return this._backgroundColor ?
+            ThemeColor.parse(
+                this._contrastService.getContrastColor(
+                    this._backgroundColor,
+                    this._lightColor,
+                    this._darkColor
+                ).toRgba())
+            : this._lightColor;
+    }
+
+    private _determineVPos(position: string): BadgeVerticalPosition {
```

Comment:
Might be good to consider how Angular Material does it's badge component: https://github.com/angular/components/blob/master/src/material/badge/badge.ts

Essentially it just adds classes to the host element: https://github.com/angular/components/blob/master/src/material/badge/badge.ts#L56

We should be able to skip most of these checks as Angular's template checking in production mode (especially in Angular 9) will throw an exception if the user tries to pass an invalid type to the property